File: backend/analytics/analytics.py
import json
import sys
import os
from datetime import datetime, timedelta
from supabase import create_client
import pandas as pd

SUPABASE_URL = os.environ.get("SUPABASE_URL")
SUPABASE_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")

def get_client():
    return create_client(SUPABASE_URL, SUPABASE_KEY)

def fetch_transactions(sb):
    data = sb.table("transactions").select("id, type, mode, digital_method, amount, party, category_id, txn_date, description, created_at").order("txn_date", desc=True).execute()
    return data.data or []

def fetch_categories(sb):
    data = sb.table("categories").select("id, name").execute()
    return {c["id"]: c["name"] for c in (data.data or [])}

def compute_summary(txns, categories):
    df = pd.DataFrame(txns)
    if df.empty:
        return {
            "overview": {"total_credit": 0, "total_debit": 0, "net_balance": 0, "txn_count": 0},
            "monthly_trend": [],
            "category_breakdown": [],
            "payment_mode_split": {"cash": {"credit": 0, "debit": 0}, "digital": {"credit": 0, "debit": 0}},
            "top_parties": [],
            "weekly_trend": [],
            "daily_avg": {"credit": 0, "debit": 0},
        }

    df["txn_date"] = pd.to_datetime(df["txn_date"])
    df["amount"] = pd.to_numeric(df["amount"], errors="coerce").fillna(0)
    df["month"] = df["txn_date"].dt.to_period("M").astype(str)
    df["week"] = df["txn_date"].dt.isocalendar().week.astype(int)
    df["weekday"] = df["txn_date"].dt.day_name()
    df["category_name"] = df["category_id"].map(categories).fillna("Uncategorized")

    total_credit = float(df[df["type"] == "credit"]["amount"].sum())
    total_debit = float(df[df["type"] == "debit"]["amount"].sum())

    # Monthly trend
    monthly = df.groupby(["month", "type"])["amount"].sum().unstack(fill_value=0).reset_index()
    monthly_trend = []
    for _, row in monthly.iterrows():
        entry = {"month": row["month"]}
        entry["credit"] = float(row.get("credit", 0))
        entry["debit"] = float(row.get("debit", 0))
        entry["net"] = entry["credit"] - entry["debit"]
        monthly_trend.append(entry)
    monthly_trend.sort(key=lambda x: x["month"])

    # Category breakdown
    cat_df = df[df["type"] == "debit"].groupby("category_name")["amount"].sum().reset_index()
    cat_df = cat_df.sort_values("amount", ascending=False)
    category_breakdown = [{"name": row["category_name"], "value": float(row["amount"])} for _, row in cat_df.iterrows()]

    # Payment mode split
    mode_df = df.groupby(["mode", "type"])["amount"].sum().unstack(fill_value=0)
    payment_mode_split = {}
    for mode in ["cash", "digital"]:
        if mode in mode_df.index:
            payment_mode_split[mode] = {
                "credit": float(mode_df.loc[mode].get("credit", 0)),
                "debit": float(mode_df.loc[mode].get("debit", 0)),
            }
        else:
            payment_mode_split[mode] = {"credit": 0, "debit": 0}

    # Top parties
    party_df = df.groupby("party")["amount"].sum().reset_index()
    party_df = party_df.dropna(subset=["party"])
    party_df = party_df.sort_values("amount", ascending=False).head(10)
    top_parties = [{"name": row["party"], "amount": float(row["amount"])} for _, row in party_df.iterrows()]

    # Weekly trend (last 7 days)
    now = df["txn_date"].max()
    week_start = now - timedelta(days=6)
    recent = df[df["txn_date"] >= week_start]
    weekly = recent.groupby([pd.Grouper(key="txn_date", freq="D"), "type"])["amount"].sum().unstack(fill_value=0).reset_index()
    weekly["date"] = weekly["txn_date"].dt.strftime("%Y-%m-%d")
    weekly_trend = []
    for _, row in weekly.iterrows():
        weekly_trend.append({
            "date": row["date"],
            "credit": float(row.get("credit", 0)),
            "debit": float(row.get("debit", 0)),
        })

    # Daily averages
    unique_days = df["txn_date"].dt.date.nunique()
    daily_avg = {
        "credit": round(total_credit / max(unique_days, 1), 2),
        "debit": round(total_debit / max(unique_days, 1), 2),
    }

    return {
        "overview": {
            "total_credit": total_credit,
            "total_debit": total_debit,
            "net_balance": total_credit - total_debit,
            "txn_count": len(df),
        },
        "monthly_trend": monthly_trend,
        "category_breakdown": category_breakdown,
        "payment_mode_split": payment_mode_split,
        "top_parties": top_parties,
        "weekly_trend": weekly_trend,
        "daily_avg": daily_avg,
    }

def main():
    sb = get_client()
    txns = fetch_transactions(sb)
    categories = fetch_categories(sb)
    result = compute_summary(txns, categories)
    print(json.dumps(result))

if __name__ == "__main__":
    main()

File: backend/src/routes/analytics.js
const express = require('express');
const supabaseAdmin = require('@/config/supabaseAdmin');
const { requireAuth } = require('@/middlewares/auth');
const { getCached, invalidate } = require('@/lib/cache');

const router = express.Router();
router.use(requireAuth);

function startOfDay(d) {
  const dt = new Date(d);
  dt.setUTCHours(0, 0, 0, 0);
  return dt;
}

function toISODate(d) {
  return d.toISOString().slice(0, 10);
}

function getMonthKey(d) {
  return d.toISOString().slice(0, 7);
}

router.get('/', async (req, res) => {
  try {
    const result = await getCached('analytics:overview', 15000, async () => {
      const { data: txns, error } = await supabaseAdmin
        .from('transactions')
        .select('id, type, mode, amount, party, category_id, txn_date, created_at');

      if (error) throw error;

      const { data: catRows } = await supabaseAdmin.from('categories').select('id, name');
      const catMap = {};
      (catRows || []).forEach(c => { catMap[c.id] = c.name; });

      const rows = (txns || []).map(t => ({
        ...t,
        _date: new Date(t.txn_date),
        _amount: parseFloat(t.amount) || 0,
        _catName: catMap[t.category_id] || 'Uncategorized',
      }));

      if (rows.length === 0) {
        return {
          overview: { total_credit: 0, total_debit: 0, net_balance: 0, txn_count: 0 },
          monthly_trend: [],
          category_breakdown: [],
          payment_mode_split: { cash: { credit: 0, debit: 0 }, digital: { credit: 0, debit: 0 } },
          top_parties: [],
          weekly_trend: [],
          daily_avg: { credit: 0, debit: 0 },
        };
      }

      let total_credit = 0;
      let total_debit = 0;

      const monthlyMap = {};
      const catMap2 = {};
      const modeMap = {};
      const partyMap = {};

      rows.forEach(r => {
        const mk = getMonthKey(r._date);
        if (!monthlyMap[mk]) monthlyMap[mk] = { month: mk, credit: 0, debit: 0, net: 0 };

        if (r.type === 'credit') {
          total_credit += r._amount;
          monthlyMap[mk].credit += r._amount;
          modeMap[r.mode] = modeMap[r.mode] || { credit: 0, debit: 0 };
          modeMap[r.mode].credit += r._amount;
        } else if (r.type === 'debit') {
          total_debit += r._amount;
          monthlyMap[mk].debit += r._amount;
          modeMap[r.mode] = modeMap[r.mode] || { credit: 0, debit: 0 };
          modeMap[r.mode].debit += r._amount;

          catMap2[r._catName] = (catMap2[r._catName] || 0) + r._amount;
        }

        if (r.party) {
          partyMap[r.party] = (partyMap[r.party] || 0) + r._amount;
        }
      });

      Object.values(monthlyMap).forEach(m => { m.net = m.credit - m.debit; });
      const monthly_trend = Object.values(monthlyMap).sort((a, b) => a.month.localeCompare(b.month));

      const category_breakdown = Object.entries(catMap2)
        .map(([name, value]) => ({ name, value }))
        .sort((a, b) => b.value - a.value);

      const payment_mode_split = {
        cash: modeMap['cash'] || { credit: 0, debit: 0 },
        digital: modeMap['digital'] || { credit: 0, debit: 0 },
      };

      const top_parties = Object.entries(partyMap)
        .map(([name, amount]) => ({ name, amount }))
        .sort((a, b) => b.amount - a.amount)
        .slice(0, 10);

      const now = rows.reduce((max, r) => r._date > max ? r._date : max, rows[0]._date);
      const weekStart = new Date(now);
      weekStart.setUTCDate(weekStart.getUTCDate() - 6);
      weekStart.setUTCHours(0, 0, 0, 0);

      const weeklyDayMap = {};
      for (let i = 0; i < 7; i++) {
        const d = new Date(weekStart);
        d.setUTCDate(d.getUTCDate() + i);
        const key = toISODate(d);
        weeklyDayMap[key] = { date: key, credit: 0, debit: 0 };
      }
      rows.forEach(r => {
        const dk = toISODate(r._date);
        if (weeklyDayMap[dk]) {
          if (r.type === 'credit') weeklyDayMap[dk].credit += r._amount;
          else weeklyDayMap[dk].debit += r._amount;
        }
      });
      const weekly_trend = Object.values(weeklyDayMap).sort((a, b) => a.date.localeCompare(b.date));

      const uniqueDays = new Set(rows.map(r => toISODate(r._date))).size || 1;
      const daily_avg = {
        credit: Math.round((total_credit / uniqueDays) * 100) / 100,
        debit: Math.round((total_debit / uniqueDays) * 100) / 100,
      };

      return {
        overview: {
          total_credit,
          total_debit,
          net_balance: total_credit - total_debit,
          txn_count: rows.length,
        },
        monthly_trend,
        category_breakdown,
        payment_mode_split,
        top_parties,
        weekly_trend,
        daily_avg,
      };
    });

    res.set('Cache-Control', 'private, max-age=15');
    res.json({ success: true, result });
  } catch (err) {
    console.error('Analytics error:', err.message);
    res.status(500).json({ success: false, message: 'Analytics computation failed' });
  }
});

module.exports = router;
